### ISSUE TITLE
Question - Matching multiple routes for a request

### ISSUE THREAD

**Issue by Dilsy99** [ThreeMammals/Ocelot#342]
## Expected Behavior

I have two routes defined like this:

```json
{
      "DownstreamPathTemplate": "/api/ClientQuery/{id}",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 52709
        }
      ],
      "UpstreamPathTemplate": "/customer/{id}",
      "UpstreamHttpMethod": [ "Get" ],
      "DelegatingHandlers": [
        "ClientQueryHandler"
      ]
    },
    {
      "DownstreamPathTemplate": "/api/ClientQuery/{id}/dependent",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 52709
        }
      ],
      "UpstreamPathTemplate": "/customer/{id}/dependent",
      "UpstreamHttpMethod": [ "Get" ],
      "DelegatingHandlers": [
        "DependentQueryHandler"
      ]
    },
```

I would expect that the call to /customer/12345/dependent would result in a call to /api/ClientQuery/12345/dependent with the DependentQueryHandler delegate handler being called

## Actual Behavior

The upstream route seems to "match" both routes defined above and the "first" route is used. This results in the call going to /api/ClientQuery/12345 and the ClientQueryHandler being used

Here is some logging info (I have added a bit of my own just to help me understand what is going on) which shows this behaviour

[15:00:03:832 Debug message: "ocelot pipeline started"
[15:00:03:847 Debug message: "Upstream url path is /customer/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19/dependent"
[15:00:03:850 Debug message: "Reroutes --> UpstreamPathTemplate: /customer/{id}"
[15:00:03:852 Debug message: "UpstreamHttpMethod: Get"
[15:00:03:854 Debug message: "Reroutes --> UpstreamPathTemplate: /customer/{id}/dependent"
[15:00:03:855 Debug message: "UpstreamHttpMethod: Get"
[15:00:03:857 Debug message: "Reroutes --> UpstreamPathTemplate: /"
[15:00:03:859 Debug message: "UpstreamHttpMethod: Get"
[15:00:03:864 Debug message: "Reroutes --> UpstreamPathTemplate: /customer"
[15:00:03:866 Debug message: "UpstreamHttpMethod: Post"
[15:00:03:868 Debug message: "Reroutes --> UpstreamPathTemplate: /customer/{id}"
[15:00:03:870 Debug message: "UpstreamHttpMethod: Put"
[15:00:03:874 Debug message: "Reroutes --> UpstreamPathTemplate: /customer/{id}/add/dependent"
[15:00:03:878 Debug message: "UpstreamHttpMethod: Post"
[15:00:03:883 Debug message: "Reroutes --> UpstreamPathTemplate: /customer/{id}/update/dependent"
[15:00:03:885 Debug message: "UpstreamHttpMethod: Post"
[15:00:03:887 Debug message: "Reroutes --> UpstreamPathTemplate: /customer/{id}/remove/dependent"
[15:00:03:890 Debug message: "UpstreamHttpMethod: Post"
[15:00:03:896 Debug message: "Upstream host: localhost:9000"
[15:00:03:901 Debug message: "applicableReRoutes --> UpstreamPathTemplate: /customer/{id} UpstreamTemplatePattern: ^(?i)/customer/[0-9a-zA-Z].*$"
[15:00:03:903 Debug message: "applicableReRoutes --> UpstreamPathTemplate: /customer/{id}/dependent UpstreamTemplatePattern: ^(?i)/customer/[0-9a-zA-Z].*/dependent$"
[15:00:03:905 Debug message: "applicableReRoutes --> UpstreamPathTemplate: / UpstreamTemplatePattern: ^/$"
[15:00:03:917 Debug message: "Path: /customer/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19/dependent"
[15:00:03:922 Debug message: "GetPlaceholderNamesAndValues -> Name: {id} Value f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19"
[15:00:03:928 Debug message: "Path: /customer/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19/dependent"
[15:00:03:934 Debug message: "GetPlaceholderNamesAndValues -> Name: {id} Value f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19"
[15:00:03:949 Debug message: "downstream templates are /api/ClientQuery/{id}"
[15:00:03:989 Information message: "EndpointRateLimiting is not enabled for /api/ClientQuery/{id}"
[15:00:04:030 Information message: "No authentication needed for /customer/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19/dependent"
[15:00:04:053 Information message: "/api/ClientQuery/{id} route does not require user to be authorised"
[15:00:04:081 Debug message: "Creating downstream url: /api/ClientQuery/{id}"
[15:00:04:083 Debug message: "Name: {id} Value f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19"
[15:00:04:090 Debug message: "Downstream url is http://localhost:52709/api/ClientQuery/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19"
[15:00:04:108 Debug message: "Cache key for request is GET:http://localhost:9000/customer/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19/dependent"
[15:00:04:161 Information Sending request to http://localhost:52709/api/ClientQuery/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19
[15:00:04:314 Information Response received from the request to http://localhost:52709/api/ClientQuery/f56bb7ef-ee42-4db1-95e1-8f0b9d9e5f19. Status received is OK
[15:00:04:318 Information HttpRequest: completed (duration: 0.17 seconds)
[15:00:04:637 Debug message: "setting http response message"
[15:00:04:659 Debug message: "no pipeline errors, setting and returning completed response"
[15:00:04:694 Debug message: "ocelot pipeline finished"
[15:00:04:698 Information BFF request received....: completed (duration: 0.90 seconds)

## Question

Looking at the code and reading the documentation, it looks as though this behavior is by design. 

Do I have any other options other than to make the upstream path templates unique (so I can get it to work by changing "/api/ClientQuery/{id}" to "/api/ClientQuery/{id}/client" for example) to make my example work as I have it defined?

Many thanks.

## Specifications

  - Version: 6.0.0
  - Platform: Windows 10
  - Subsystem: Visual Studio 2017 Net Core 2.0

**Comment by juancash** [comment]
Hi @Dilsy99, it seems to be by design. 
You must write the most specific rules first and the most ambiguous rules at the end.

**Comment by Dilsy99** [comment]
Thank you both @juancash  @TomPallister.

I think I was thinking about it in a more as a "one to one" relationship with the configured reroutes rather than a lookup based on all that are defined. Shows where assumptions get you!

The priority tweak in config works so thank you for that.

**Comment by TomPallister** [comment]
@Dilsy99 how bad do you think this logic is? I think it gives you a lot of power because you can configure anything but it is complicated and confusing e.g. you could maybe have a specific wildcard character in the ReRoute!

**Comment by Dilsy99** [comment]
Hi @TomPallister.

In my simple mind I had it more as a 1 to 1 relationship with each reroute. Having looked at the code (and re-read the docs :-)) I see how the current logic provides a lot of flexibility (arguably at the cost of being a little opaque to the consumer).

I did wonder about more of a "ranked" search rather than first wins (sort of like the priority but baked in). 
Simplistically, it could match number an equal number of "segments" in the request higher than not, a closer number of "segments" in the request higher than not etc. Equal ranks could then be based on order in the reroute. Not sure if that would break the flexibility but was a thought!

The other way I suppose, is to be able to inject in your own "find the route" behavior. Looking at the code, being able to inject in a custom UrlPathToUrlTemplateMatcher type class into DownstreamRouteFinder might provide this. Probably would provide the most flexibility for the users but probably the most work to implement!

Cheers.

**Comment by TomPallister** [comment]
@Dilsy99 thanks for your comment, I will keep this on my todo list and when I've got all the other issues sorted take a look. I really appreciate the input, this project will not be good without it.

**Comment by Dilsy99** [comment]
@TomPallister 
It is good to be involved!
If it is something that I can help with I can probably get time from within the project I am working on at the moment. I am hoping that we end up using Ocelot (we are at POC stage at the moment) so anything that makes it work better for us (and hopefully other people) is something I think I can sell!
Cheers

**Comment by TomPallister** [comment]
@Dilsy99 Interesting! What other API gateways are you looking at?

If you want you could have a playaround with the route matching stuff? I'm introducing a factory for the DownstreamRouteFinderMiddleware so you can match routes in different ways for another ticket. This might be useful.

**Comment by Dilsy99** [comment]
@TomPallister To be honest, Ocelot or do it ourselves at the moment!!!

If you want to suggest another one I am all ears :-)

We started using Ocelot because it allowed me to get a POC up and running nice a quickly and seemed to provide the functions we thought we may need. In fact, the "features" page on the documentation was almost a tick list of things we were going to need to do!!

It has worked well so far and we have found a way to do almost everything we have needed (so far ;-)). One of our main aims was to prove we could use a "Backends For Frontends" (BFF) to augment our microservices responses with Hateoas links to drive our UI. We have used the delegate handlers linked to the routes to do this and we have been able to start driving the UI. This also allowed us to "massage" the responses in other ways which has proved useful. We will refactor how we have implemented our handler code but for the POC it stands up!

Things we are also looking at dealing with are:
Security
State management
Response aggregation

I think a factory for DownstreamRouteFinderMiddleware sounds exactly like what we need. I will keep my eyes open for that feature.

Cheers
